Zephyr modules: CMake and Kconfig build integration

tejlmand opened the pull request:
This commit allows for Zephyr modules to be natively integrated into
the build system with CMakeLists.txt and Kconfig files.

The sourcing of module files are done in following order:
- If <module>/zephyr/module.yml exists, use cmake and kconfig settings
  for sourcing of additional file
- Else if <module>/zephyr/CMakeLists.txt exists, source this file into
  CMake build tree and add <module>/zephyr/Kconfig as osource

If none of the above files are present, the project is considered to
not be a Zephyr module

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

carlescufi committed:
doc: west: Document build system integration

Document the build system integration with west and any projects managed
by it.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

tejlmand committed:
CMake/Kconfig: CMake and Kconfig build integration

This commit allows for Zephyr modules to be natively integrated into
the build system with CMakeLists.txt and Kconfig files.

The sourcing of module files are done in following order:
- If <module>/zephyr/module.yml exists, use cmake and kconfig settings
  for sourcing of additional file
- Else if <module>/zephyr/CMakeLists.txt exists, source this file into
  CMake build tree and add <module>/zephyr/Kconfig as osource

If none of the above files are present, the project is considered to
not be a Zephyr module

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

mped-oticon commented:
@tejlmand From the title and commit message I could not see that this was meant for _west_ modules.

mped-oticon commented on scripts/yaml_to_cmake.py:
```diff
@@ -0,0 +1,68 @@
+# SPDX-License-Identifier: Apache-2.0
+
+'''Tool for a simple parsing of YAML files and return a ;-list of <key>=<value>
+pairs to use within a CMake build file.
```
but this is still only meant for use with west, right?

mbolivar replied:
> but this is still only meant for use with west, right?

The idea is that users who do not want to use west to manage projects can set ZEPHYR_MODULES directly at CMake time.

tejlmand replied:
This script is intended to be a general way of converting a yaml file containing key-value maps into a `;`-list that can be used by CMake.
Right now it can convert a `module.yml` file with the embedded schema, but there is no reason it cannot be extended to take a schema-file as argument, and thus convert an yaml-file.

mbolivar commented on scripts/yaml_to_cmake.py:
```diff
@@ -0,0 +1,68 @@
+## A pykwalify schema for basic validation of the structure of a
+## metadata YAML file.
+##
+# The zephyr/module.yml file is a simple list of key value pairs to be used in
```
s/ in// (can delete "in")

tejlmand replied:
done

mbolivar reviewed (commented):
Looks good to me -- consistent with the approach we have been discussing.

tejlmand commented:
> @tejlmand From the title and commit message I could not see that this was meant for _west_ modules.

@mped-oticon This is not `west` modules. It is a principle for having modules for Zephyr build system that can be located in independent repos (Out of Zephyr tree).
West can be used to fetch those repos, and if west is available, then it will be used to get a list of repos.
But user can also specify those modules using `cmake -DZEPHYR_MODULES=<path-to-module>[;<path-to-additional-module(s)>]` if not using west for repo management.

tejlmand commented on cmake/zephyr_module.cmake:
```diff
@@ -0,0 +1,70 @@
+endif()
+
+# Clear the Kconfig.modules generated file in case modules has been removed.
+file(WRITE ${PROJECT_BINARY_DIR}/Kconfig.modules "")
```
Good point.
Done

carlescufi replied:
I am not so sure about `modules.generated.kconfig`. We currently don't follow that naming scheme, examples:

`autoconf.h`
`offsets.h`
`generated_dts_board.conf`

I would be in favor of something like `Kconfig.generated.modules` though.

tejlmand replied:
Just to be clear, this is what I did:
>     * Add a header of "# NOTE: THIS IS AUTOGENERATED BY " to this file.

tejlmand replied:
File is after all located inside the build folder.
Which imho indicates that it is created and used by the build system.

tejlmand commented on cmake/zephyr_module.cmake:
```diff
@@ -0,0 +1,70 @@
+## Clear zephyr modules property
```
Absolutely.
Done.

tejlmand commented:
Made a small update to the yml schema.
Removed the sequence, so that a valid `zephyr/module.yml` now looks as:
```
build:
  cmake: .
  kconfig: zephyr/Kconfig
```
that is without `-` in front of keys.

carlescufi commented:
@mped-oticon I now pushed a commit with the doc, that should clarify things a bit

carlescufi commented on cmake/zephyr_module.cmake:
```diff
@@ -0,0 +1,83 @@
+# Clear the Kconfig.modules generated file in case modules has been removed.
+# The Kconfig.modules contains a list of additional Kconfig files to be sourced
+# based upon <module>/zephyr/module.yml files.
+file(WRITE ${PROJECT_BINARY_DIR}/Kconfig.modules
```
What is `PROJECT_BINARY_DIR`? is it `build/zephyr`?

mped-oticon replied:
Yes.

tejlmand replied:
And that is already used else-where in Kconfig.
So I decided to use same property in this PR, instead of exporting yet another property.

carlescufi commented on cmake/zephyr_module.cmake:
```diff
@@ -0,0 +1,83 @@
+          )
+      endif()
+    endforeach()
+
```
Can we print a message here:
`Including modules: tinycbor, mbedtls, ...`

carlescufi commented on CMakeLists.txt:
```diff
@@ -551,6 +551,16 @@ add_subdirectory(subsys)
 add_subdirectory(drivers)
 add_subdirectory(tests)
 
+# Add all zephyr modules subdirectories.
```
Can we print a message here:
`Including modules: tinycbor, mbedtls, ...`

tejlmand replied:
Well, those are not zephyr modules in `west.yml` yet.
Will they become that prior to 1.14 ?

mbolivar replied:
I think @carlescufi means "print the list of modules discovered at runtime", not that exact list

tejlmand replied:
Ah, got it.
Good point, will do that.

tejlmand replied:
Included the print message a couple of lines below.

tejlmand commented on doc/application/index.rst:
```diff
@@ -641,6 +641,73 @@ Zephyr binary into your application directory.
+
+  - If a ``cmake: <folder>`` field is present then the build system will match
+    the :file:`CMakeLists.txt` in ``<folder>``
+  - If a ``kconfig: <folder>`` field is present then the build system will match
```
note: For kconfig, folder is not enough.
`module.yml` must provide full name, i.e. 
```
  kconfig: <folder>/<kconfig-file>
```
as the kconfig file can have any name, whereas, cmake requires `CMakeLists.txt` when using `add_subdirectory`

carlescufi replied:
Understood, can you please fix that along with @dbkinder's suggestions?

tejlmand replied:
done

tejlmand commented on doc/application/index.rst:
```diff
@@ -641,6 +641,73 @@ Zephyr binary into your application directory.
+  :file:`<path>/zephyr/CMakeLists.txt` the build system will match it
+- If the project contains a file named :file:`<path>/zephyr/Kconfig` the build
+  system will match it
+
```
Should we present a sample `module.yml` file here ?
As example:
```
build:
  cmake: .
  kconfig: Kconfig
```

carlescufi replied:
Yes, let's show a simple example here

tejlmand replied:
done

dbkinder commented on doc/application/index.rst:
```diff
@@ -129,8 +129,8 @@ subdirectories which are not described here.
-    from other sources, such as hardware interface code supplied by
-    manufacturers and cryptographic library code.
+    from other sources and that must live inside the zephyr repository unlike
+    `external projects <ext-projs>`
```
Needs a trailing underscore to turn into a link:
```suggestion
    `external projects <ext-projs>`_
```

dbkinder commented on doc/application/index.rst:
```diff
@@ -641,6 +641,73 @@ Zephyr binary into your application directory.
+mature code as possible when it makes sense. In the context of Zephyr's build
+system those are called *modules*.
+There are several categories of external projects that Zephyr depends on,
+among which:
```
```suggestion
including:
```

dbkinder commented on doc/application/index.rst:
```diff
@@ -641,6 +641,73 @@ Zephyr binary into your application directory.
+
+The way that the build system discovers and includes external projects when
+building an application is based around the :ref:`west` tool's ``west list``
+command. The main CMake script in zephyr (the :file:`CMakeLists.txt` located
```
Capitalize ``Zephyr`` when it's the project/OS name...

dbkinder commented on doc/application/index.rst:
```diff
@@ -641,6 +641,73 @@ Zephyr binary into your application directory.
+building an application is based around the :ref:`west` tool's ``west list``
+command. The main CMake script in zephyr (the :file:`CMakeLists.txt` located
+in the zephyr repository root folder) invokes ``west list`` to obtain a list of
+projects managed by west. For each project then the CMake scripts will verify
```
```suggestion
projects managed by west. For each project then, the CMake scripts will verify
```

dbkinder commented on doc/application/index.rst:
```diff
@@ -641,6 +641,73 @@ Zephyr binary into your application directory.
+command. The main CMake script in zephyr (the :file:`CMakeLists.txt` located
+in the zephyr repository root folder) invokes ``west list`` to obtain a list of
+projects managed by west. For each project then the CMake scripts will verify
+if it contains the required metadata required in a module and, if it does,
```
```suggestion
if it contains the metadata required in a module.  If it does,
```

dbkinder commented on doc/application/index.rst:
```diff
@@ -641,6 +641,73 @@ Zephyr binary into your application directory.
+in the zephyr repository root folder) invokes ``west list`` to obtain a list of
+projects managed by west. For each project then the CMake scripts will verify
+if it contains the required metadata required in a module and, if it does,
+it will include it in the build.
```
With three ``it``s in this sentence, I'm not sure what the last ``it`` is referring to here.  If it's a module, then say that:  ``... it will include that module in the build.``

tejlmand replied:
Tried to rephrase to make the sentence clearer.

SebastianBoe commented on cmake/zephyr_module.cmake:
```diff
@@ -0,0 +1,83 @@
+    OUTPUT_VARIABLE
+    west_project_list
+  )
+  string(REGEX REPLACE "[\r\n]+" ";" west_project_list "${west_project_list}")
```
Please don't mutate variables.

So, for instance

`string(REGEX REPLACE "[\r\n]+" ";" west_project_list "${west_list_output}")`

tejlmand replied:
done

SebastianBoe commented on Kconfig.zephyr:
```diff
@@ -40,6 +40,8 @@ source "ext/Kconfig"
 
 source "tests/Kconfig"
 
+osource "$(PROJECT_BINARY_DIR)/Kconfig.modules"
```
Kconfig.modules is not optional, so

`source "$(PROJECT_BINARY_DIR)/Kconfig.modules"`

tejlmand replied:
in principle we could decide not to generate the file if no modules are present.
But as we always create the file (but without content if there are no modules), then I adopted your suggestion.

SebastianBoe commented on scripts/yaml_to_cmake.py:
```diff
@@ -0,0 +1,65 @@
+    val_str = ''
+    for key in section:
+        val_str += '{}={};'.format(key, section[key])
+    print(val_str)
```
In Zephyr it is convention to use output files instead of STDOUT.

Please align.

SebastianBoe replied:
After aligning, --file is ambiguous, so you need;

`-i, --input`
`-o, --output`

tejlmand replied:
will do

tejlmand replied:
> In Zephyr it is convention to use output files instead of STDOUT.

Just looked around.
Are you sure you want this output to go to a file ?
When files are written in other places, it's because they are to be used later during the build, e.g. generation of header files.
This data is supposed to be used INSIDE CMake, during configure time.
Thus writing the data to a file will result in additional code inside CMake, such as
```
FILE(READ <converted-data>)
```
instead of getting this data directly into a variable as is done now.
If think you miss the point of this code.
It's purpose is to take an existing `module.yml` and then has the content available inside CMake as a list.

I changed `--file` to `-i | --input`

tejlmand replied:
Fixed.
now using `[-i|--output]` and [-o|--output] instead.

SebastianBoe commented on scripts/yaml_to_cmake.py:
```diff
@@ -0,0 +1,65 @@
+    for key in section:
+        val_str += '{}={};'.format(key, section[key])
+    print(val_str)
+    sys.exit(0)
```
This is non-standard, I believe it can be safely dropped.

tejlmand replied:
done

SebastianBoe commented on scripts/yaml_to_cmake.py:
```diff
@@ -0,0 +1,65 @@
+
+def main():
+    parser = argparse.ArgumentParser(description='''
+    YAML to CMake variable converter''')
```
Converts YAML to a CMake list.

tejlmand replied:
done

SebastianBoe commented on cmake/zephyr_module.cmake:
```diff
@@ -0,0 +1,83 @@
+    # Not a Zephyr module, ignore.
+  endif()
+
+  file(APPEND ${PROJECT_BINARY_DIR}/Kconfig.modules ${kconfig_osource})
```
What happens when kconfig_osource is set in one loop iteration, and then not set in the next,
won't you get duplicate entries?

Or does foreach create a new scope? I thought only functions and directories created new scopes ...

tejlmand replied:
nope, you're correct.
Thanks for spotting.

SebastianBoe commented on cmake/zephyr_module.cmake:
```diff
@@ -0,0 +1,83 @@
+    set_property(GLOBAL APPEND PROPERTY ZEPHYR_MODULES_NAME ${module_name})
+    set_property(GLOBAL APPEND PROPERTY ZEPHYR_MODULES_DIR ${module}/${cmake_subdir})
+  elseif(EXISTS "${module}/${cmake_subdir}/CMakeLists.txt")
+    set(kconfig_osource "${module}/zephyr/Kconfig\n")
```
shouldn't it be 

`osource ${module}/zephyr/Kconfig\n`

?

tejlmand replied:
correct

SebastianBoe commented on cmake/zephyr_module.cmake:
```diff
@@ -0,0 +1,83 @@
+# search for zephyr/module.yml
+
+# Clear zephyr modules property
+set_property(GLOBAL PROPERTY ZEPHYR_MODULES_NAME "")
```
For "minimal possible scope" reasons, symbols should be variables until a need for more scope is proven.

tejlmand replied:
Fixed.
Actually no need for a property here.

SebastianBoe commented on cmake/zephyr_module.cmake:
```diff
@@ -0,0 +1,83 @@
+      if(${key_value} MATCHES "^cmake=")
+        string(REGEX REPLACE "^cmake=" "" cmake_subdir ${key_value})
+      endif()
+      if(${key_value} MATCHES "^kconfig=")
```
elseif

tejlmand replied:
changed

carlescufi commented on cmake/zephyr_module.cmake:
```diff
@@ -0,0 +1,83 @@
+  ## Use `west list` to fetch all west handled projects.
+  execute_process(
+    COMMAND
+    west list --format={abspath}\;
```
ah, we are already looking for west using `find_program()`, so this just needs to use it here

carlescufi replied:
And in fact this should be optional, running only if `WEST` is defined

tejlmand replied:
Fixed.

tejlmand commented on cmake/zephyr_module.cmake:
```diff
@@ -0,0 +1,83 @@
+  ## Use `west list` to fetch all west handled projects.
+  execute_process(
+    COMMAND
+    west list --format={abspath}\;
```
it would place a `;` after each project, unfortunately the way the list is printed we still need to remove new lines.
Thus we can remove the `\;` because the regex below.

nashif commented:
recheck

SebastianBoe commented on cmake/zephyr_module.cmake:
```diff
@@ -0,0 +1,84 @@
+      )
+    file(
+      STRINGS
+      "${CMAKE_CURRENT_BINARY_DIR}/zephyr_module_data.txt"
```
Since the file is mutated, we will add duplicates when "section is None".

Making the file immutable would solve this issue (new file for each module).

tejlmand replied:
changed into always writing the file, even if empty, as I think that is actually the most correct thing to do.
Empty section in yaml -> no data in converted file

tejlmand replied:
Also made new file for each module.
This will help debugging if need be, as well as in general looks nicer.

carlescufi reviewed (approved):
looks good, thanks @tejlmand!

mbolivar commented:
@tejlmand as you requested on slack, I have force-pushed a change to use posixpath instead of abspath as the west list format key.

tejlmand commented:
> @tejlmand as you requested on slack, I have force-pushed a change to use posixpath instead of abspath as the west list format key.

@mbolivar Thank you very much for the help

nashif commented on Kconfig.zephyr:
```diff
@@ -40,6 +40,8 @@ source "ext/Kconfig"
 
 source "tests/Kconfig"
 
+source "$(PROJECT_BINARY_DIR)/Kconfig.modules"
```
this is breaking documentation and kconfig CI

nashif replied:
I fixed that now

nashif committed:
doc: create an empty Kconfig.modules

This file is now needed by Kconfig and was not created during doc
creation. So touch an empty file to make doc build go through.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
